networking_calico/plugins/ml2/drivers/calico/policy.py

```diff
@@ -140,17 +140,16 @@ def _neutron_rule_to_etcd_rule(rule):
     if rule['protocol'] is None or rule['protocol'] == -1:
         pass
     elif rule['protocol'] == 'ipv6-icmp':
-        etcd_rule['protocol'] = 'ICMPv6'
+        etcd_rule['protocol'] = {'IPv6': 'ICMPv6'}[ethertype]
```

Comment:
I don't think this change is useful, and it looks more complicated.  Unless it really achieves that I'm missing, please revert.

Reply:
It is indeed required. Else it will break custom icmp type/code rules. It is not required for the all icmp rule.

Reply:
Sorry, I don't quite understand yet.  If ethertype is 'IPv6', this gives 'ICMPv6' as before.  If rule['protocol'] is 'ipv6-icmp' but ethertype is not 'IPv6', does that have a valid meaning, or is it just an invalid combination?

Reply:
I would say it's an invalid combination. All ipv6 packets must have the correct ipv6 ethertype. All I can see is that the etcd key cannot be parsed on the compute node and no iptables filter is created when the [ethertype] is missing in the code.
I haven't learnt how to peek into etcd yet.

Reply:
Now I have learnt how to peek. If ethertype is missing it's not matched as icmp, ie, `rule['protocol'] == 'ipv6-icmp'`
<pre>
OK
{"action": "Allow", "ipVersion": 6, "icmp": {"code": 0, "type": 128}, "protocol": "ICMPv6", "source": {"nets": ["::/0"]}}
NOK
{"action": "Allow", "ipVersion": 6, "destination": {"ports": ["128:0"]}, "protocol": "IPV6-ICMP", "source": {"nets": ["::/0"]}}
</pre>

Reply:
I'm sorry, I still don't get this.  What specifically do you think is wrong with `etcd_rule['protocol'] = 'ICMPv6'`, and how (in detail) do you think the behaviour will be better with `etcd_rule['protocol'] = {'IPv6': 'ICMPv6'}[ethertype]`?

Reply:
Simply because it doesn't work without it. You cannot make an icmp-v6 type/code rule without it. You also have it on the original "icmp" condition so why not have the same coding on this additional condition. (The part I reverted for compability)
I'm not a python guy so I can't really say why `rule['protocol'] == 'ipv6-icmp'` wouldn't be equal without it but it's how it plays out. You end up with `{"ports": ["128:0"]}` instead of `{"code": 0, "type": 128}` which breaks the whole rule.

Reply:
Now I have learnt some python too and you are right. My code looks up the right value in the hash but since it's only one choice in the hash it is better to set it directly. I don't know why it failed for me without it before. I've tested it again and it works as expected. I will update the PR.